src/utilities/sql/PreparedStatement.hpp

```diff
@@ -0,0 +1,315 @@
+          + ", extended code " + std::to_string(extendedErrorCode) + ", errmsg: " + errMsg);
+    }
+  }
+
+  template<typename... Args>
+  PreparedStatement(const std::string &t_stmt, sqlite3 *t_db, bool t_transaction = false, Args&&... args)
+  : m_db(t_db), m_statement(nullptr), m_transaction(t_transaction)
+  {
+    if (m_transaction)
+    {
+      sqlite3_exec(m_db, "BEGIN", nullptr, nullptr, nullptr);
+    }
+
+    int code = sqlite3_prepare_v2(m_db, t_stmt.c_str(), t_stmt.size(), &m_statement, nullptr);
+
+    if (!m_statement)
+    {
+      int extendedErrorCode = sqlite3_extended_errcode(m_db);
+      const char * err = sqlite3_errmsg(m_db);
+      std::string errMsg = err;
+      throw std::runtime_error("Error creating prepared statement: " + t_stmt + " with error code " + std::to_string(code)
+          + ", extended code " + std::to_string(extendedErrorCode) + ", errmsg: " + errMsg);
+    }
+
+    if (!bindAll(args...)) {
+      throw std::runtime_error("Error bindings args with statement: " + t_stmt);
+    }
+  }
+
+  ~PreparedStatement()
+  {
+    if (m_statement)
+    {
+      sqlite3_finalize(m_statement);
+    }
+
+    if (m_transaction)
+    {
+      sqlite3_exec(m_db, "COMMIT", nullptr, nullptr, nullptr);
+    }
+  }
+
+  bool bind(int position, const std::string &t_str)
+  {
+    if (sqlite3_bind_text(m_statement, position, t_str.c_str(), t_str.size(), SQLITE_TRANSIENT) != SQLITE_OK) {
+      return false;
+    }
+    return true;
+  }
+
+  bool bind(int position, int val)
+  {
+    if (sqlite3_bind_int(m_statement, position, val) != SQLITE_OK) {
+      return false;
+    }
+    return true;
+  }
+
+  bool bind(int position, unsigned int val)
+  {
+    if (sqlite3_bind_int(m_statement, position, val) != SQLITE_OK) {
+      return false;
+    }
+    return true;
+  }
+
+
+  bool bind(int position, double val)
+  {
+    if (sqlite3_bind_double(m_statement, position, val) != SQLITE_OK) {
+      return false;
+    }
+    return true;
+  }
+
+  template<typename... Args>
+  bool bindAll(Args&&... args) {
+    const std::size_t size = sizeof...(Args);
+    int nPlaceholders = sqlite3_bind_parameter_count(m_statement);
+    if (nPlaceholders != size) {
+      throw std::runtime_error("Wrong number of placeholders [" + std::to_string(nPlaceholders) + "] versus bindArgs [" + std::to_string(size) + "].");
+      return false;
+    } else {
+      int i = 1;
+      // Call bind(i++, args[i]) until any returns false
+      bool ok = (bind(i++, args) && ...);
+      return ok;
+    }
+  }
+
+  // Executes a **SINGLE** statement
+  void execAndThrowOnError() {
+    int code = execute();
+    if (code != SQLITE_DONE)
+    {
+      throw std::runtime_error("Error executing SQL statement step");
+    }
+  }
+
+  // Executes a **SINGLE** statement
+  int execute() {
+    int code = sqlite3_step(m_statement);
+    sqlite3_reset(m_statement);
+    return code;
+  }
```

Comment:
The non variadic functions could be put into a cpp file, but I didn't do it yet.